core/src/main/scala/org/apache/spark/h2o/H2OContext.scala

```diff
@@ -255,7 +265,142 @@ object H2OContext extends Logging{
     * @return H2O Context
     */
   def getOrCreate(sc: SparkContext): H2OContext = {
-    getOrCreate(sc, new H2OConf(sc))
+    getOrCreate(sc, new H2OConf(sc))(SQLContext.getOrCreate(sc))
+  }
+
+  val defaultFieldNames = (i: Int) => "f" + i
+
+  def buildH2OFrame(kn: String, vecTypes: Array[Byte], res: Array[Long]): H2OFrame = {
+    val u = this.asInstanceOf[ConverterUtils]
+    new H2OFrame(u.finalizeFrame(kn, res, vecTypes))
+  }
+
+  case class MetaInfo(names:Array[String], vecTypes: Array[Byte]) {
+    require(names.length > 0, "Empty meta info does not make sense")
+    require(names.length == vecTypes.length, s"Different lengths: ${names.length} names, ${vecTypes.length} types")
+  }
+
+  private
+  def initFrame[T](keyName: String, names: Array[String]):Unit = {
+    val fr = new water.fvec.Frame(Key.make(keyName))
+    water.fvec.FrameUtils.preparePartialFrame(fr, names)
+    // Save it directly to DKV
+    fr.update()
+  }
+
+  //  private
+  //  def finalizeFrame[T](keyName: String,
+  //                       res: Array[Long],
+  //                       colTypes: Array[Byte],
+  //                       colDomains: Array[Array[String]] = null):Frame = {
+  //    val fr:Frame = DKV.get(keyName).get.asInstanceOf[Frame]
+  //    water.fvec.FrameUtils.finalizePartialFrame(fr, res, colDomains, colTypes)
+  //    fr
+  //  }
+
+  def keyName(rdd: RDD[_], frameKeyName: Option[String]) = frameKeyName.getOrElse("frame_rdd_" + rdd.id + Key.rand())
+
+  case class FromPureProduct(sc: SparkContext, rdd: RDD[Product], frameKeyName: Option[String]) {
+
+    import scala.reflect.runtime.universe._
+
+    def withDefaultFieldNames() = {
+      withFieldNames(defaultFieldNames)
+    }
+
+    def withFieldNames(fieldNames: Int => String): H2OFrame = {
+      val meta = metaInfo(fieldNames)
+      withMeta(meta)
+    }
+
+    def withFields(tuples: List[(String, Type)]): H2OFrame = {
+      val meta = metaInfo(tuples)
+      withMeta(meta)
+    }
+
+    def withMeta(meta: MetaInfo): H2OFrame = {
+
+      val kn: String = keyName(rdd, frameKeyName)
+      // Make an H2O data Frame - but with no backing data (yet)
+      initFrame(kn, meta.names)
+      // Create chunks on remote nodes
+      val rows = sc.runJob(rdd, partitionJob(kn, meta.vecTypes)) // eager, not lazy, evaluation
+      val res = new Array[Long](rdd.partitions.length)
+      rows.foreach { case (cidx, nrows) => res(cidx) = nrows }
+
+      // Add Vec headers per-Chunk, and finalize the H2O Frame
+      buildH2OFrame(kn, meta.vecTypes, res)
+    }
+
+    private
+    def perTypedDataPartition[A<:Product](keystr:String, vecTypes: Array[Byte])
+                                         ( context: TaskContext, it: Iterator[A] ): (Int,Long) = {
+      // An array of H2O NewChunks; A place to record all the data in this partition
+      val nchks = water.fvec.FrameUtils.createNewChunks(keystr, vecTypes, context.partitionId)
```

Comment:
I would it it as part of this commit since you created this code as part of this change and also otherwise we would have too kinds of different APIs at the same time. Code for transforming is already not in H2OContext so proper integration with current api would be to put it into converters and unify it wit the existent api.

From my point of view making the code to play nicely with the rest of the code is also part of any change, even though the primary goal of that change is different.